pkg/controller/vitessshard/reconcile_disk.go

```diff
@@ -0,0 +1,129 @@
+package vitessshard
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+)
+
+const (
+	pvcDiskSizeAnnotation = "planetscale.com/pvc-filesystem-resize"
+)
+
+func (r *ReconcileVitessShard) reconcileDisk(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+	anythingChanged := false
+	tabletPods, err := r.tabletPodsFromShard(ctx, vts)
+	if err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	for i  := range vts.Spec.TabletPools {
+		tabletPool := &vts.Spec.TabletPools[i]
+		requestedDiskQuantity := tabletPool.DataVolumeClaimTemplate.Resources.Requests[v1.ResourceStorage]
+		requestedDisk := requestedDiskQuantity.String()
+		poolTablets, err := tabletsForPool(vts, tabletPool.Cell, string(tabletPool.Type))
+		if err != nil {
+			return resultBuilder.Error(err)
+		}
+
+		for _, tabletKey := range poolTablets {
+			pod, ok := tabletPods[tabletKey]
+			if !ok {
+				continue
+			}
+
+			pvc, err := r.claimForTabletPod(ctx, pod)
+			if err != nil {
+				return resultBuilder.Error(err)
+			}
+
+			// If the PVC does not have the FileSystemResizeCondition, bail out.
+			if !checkPVCFileSystemResizeCondition(pvc) {
+				return resultBuilder.Result()
+			}
+
+			// If we have reached this point in the loop, it indicates that there are disk size changes, so we
+			// set the variable anythingChanged to true. If we successfully complete this loop without bailing,
+			// we can be certain that we have disk size changes and that all required changes have been set.
+			anythingChanged = true
+
+			// If the pod does not have updates scheduled, bail out and wait until the scheduled annotation is applied.
+			if !rollout.Scheduled(pod) {
+				return resultBuilder.Result()
+			}
+
+			// Finally, if the PVC's disk spec does not equal the new requested disk, bail out.
```

Comment:
We should check this before we check the resize condition. The condition means "I'm ready to resize", but it might be ready to resize to an old value, which is why we do this additional check.

In general, before we check whether some goal is reached, we should confirm that the goal itself is what we think it should be.